python/cudf/cudf/tests/test_spilling.py

```diff
@@ -680,9 +696,59 @@ def test_statistics_expose(manager: SpillManager):
 
     # Expose the new buffers and check that they are counted as spilled
     for i in range(10):
-        buffers[i].mark_exposed()
+        buffers[i].owner.mark_exposed()
     assert len(manager.statistics.exposes) == 3
     stat = list(manager.statistics.exposes.values())[2]
     assert stat.count == 10
     assert stat.total_nbytes == buffers[0].nbytes * 10
     assert stat.spilled_nbytes == buffers[0].nbytes * 10
+
+
+def test_spill_on_demand(manager: SpillManager):
+    with set_rmm_memory_pool(1024):
+        a = as_buffer(data=rmm.DeviceBuffer(size=1024))
+        assert isinstance(a, SpillableBuffer)
+        assert not a.is_spilled
+
+        with pytest.raises(MemoryError, match="Maximum pool size exceeded"):
+            as_buffer(data=rmm.DeviceBuffer(size=1024))
+
+        with spill_on_demand_globally():
+            b = as_buffer(data=rmm.DeviceBuffer(size=1024))
+            assert a.is_spilled
+            assert not b.is_spilled
+
+        with pytest.raises(MemoryError, match="Maximum pool size exceeded"):
+            as_buffer(data=rmm.DeviceBuffer(size=1024))
+
+
+def test_spilling_and_copy_on_write(manager: SpillManager):
```

Comment:
@galipremsagar are there more tests that you think we should add here for the case where spilling and copy-on-write are enabled simultaneously? Right now it's mostly testing the behavior of spilling when copy-on-write is enabled, but we might want to also add some testing of when we expect copies to be made by copy-on-write in case having spilling active affects that (ideally it won't unless we exceed the spill limits, but we should probably validate that such cases behave sensibly).

Reply:
Added some more tests: https://github.com/rapidsai/cudf/pull/15436/commits/6fe2d58fb195a129cc74502155a253ff00898ed1

Reply:
I'm happy with the current state, but would like to hear Prem's take before we merge since he has the most familiarity with c-o-w.

Reply:
Looks good to me 👍